libsolidity/ExpressionCompiler.cpp

```diff
@@ -1064,36 +1089,60 @@ void ExpressionCompiler::LValue::storeValue(Expression const& _expression, Type
 		}
 		else
 		{
-			solAssert(!_move, "Move assign for non-value types not implemented.");
-			solAssert(_sourceType.getCategory() == _expression.getType()->getCategory(), "");
-			if (_expression.getType()->getCategory() == Type::Category::ByteArray)
+			solAssert(_sourceType.getCategory() == m_dataType->getCategory(), "");
+			if (m_dataType->getCategory() == Type::Category::ByteArray)
 				CompilerUtils(*m_context).copyByteArrayToStorage(
-							dynamic_cast<ByteArrayType const&>(*_expression.getType()),
+							dynamic_cast<ByteArrayType const&>(*m_dataType),
 							dynamic_cast<ByteArrayType const&>(_sourceType));
-			else if (_expression.getType()->getCategory() == Type::Category::Struct)
+			else if (m_dataType->getCategory() == Type::Category::Struct)
 			{
-				//@todo
-				solAssert(false, "Struct copy not yet implemented.");
+				// stack layout: source_ref target_ref
+				auto const& structType = dynamic_cast<StructType const&>(*m_dataType);
```

Comment:
When you do a dynamic cast to StructType you do not need auto, because you know exactly which type it will be.
Can you explain why you need that dynamic cast?